src/Steps/Release/PublishRelease.php

```diff
@@ -288,6 +290,132 @@ protected function updateGithubChangelog(OutputInterface $output, LibraryRelease
+    }
+
+    /**
+     * Get the name of the tag to use as the previous version for generating release notes
+     */
+    private function getPreviousReleaseTag(
+        OutputInterface $output,
+        LibraryRelease $release
+    ): string {
+        $version = $release->getVersion();
+        $major = $version->getMajor();
+        $minor = $version->getMinor();
+        $patch = $version->getPatch();
+
+        // Patch release (e.g. 1.2.3)
+        // Previous tag is x.y.z-1 (e.g. 1.2.2)
+        if ($patch != 0) {
+            $previousPatch = $patch - 1;
+            $previousTag = "$major.$minor.$previousPatch";
+            if ($output->isVeryVerbose()) {
+                $this->log($output, "Release is a patch. Previous tag is $previousTag");
+            }
+            return $previousTag;
+        }
+
+        // Minor release (e.g. 1.3.0)
+        // Previous tag is x.y-1.* where * is the latest patch available (e.g. 1.2.3)
+        if ($minor != 0) {
+            $previousMinor = $minor - 1;
+            // Find most recent tag for previous minor
+            $previousTag = $this->discoverMostRecentTag($major, $previousMinor, $release->getLibrary()->getTags());
+            if (!$previousTag) {
+                // If we couldn't find the tag to use as the previous version, just let GitHub infer it
+                $this->log($output, 'Could not find a reliable latest tag for the previous minor');
+            } else if ($output->isVeryVerbose()) {
+                $this->log($output, "Release is a minor. Previous tag is $previousTag");
+            }
+            return $previousTag;
+        }
+
+        // Major release (e.g. 2.0.0)
+        // Previous tag is x-1.*.* where * is the latest minor/patch available (e.g. 1.2.3)
+        $previousMajor = $major - 1;
+        if ($previousMajor != 0) {
+            # Find most recent tag for previous major
+            $previousTag = $this->discoverMostRecentTag($previousMajor, null, $release->getLibrary()->getTags());
+            if (!$previousTag) {
+                // If we couldn't find the tag to use as the previous version, just let GitHub infer it
+                $this->log($output, 'Could not find a reliable latest tag for the previous major');
+            } else if ($output->isVeryVerbose()) {
+                $this->log($output, "Release is a major. Previous tag is $previousTag");
+            }
+            return $previousTag;
+        }
+
+        // For major 1.0.0 releases, there is no suitable previous version to use.
+        $this->log($output, 'This is the first stable major version. No suitable previous tag to choose from');
+        return '';
+    }
+
+    /**
+     * Discover the most recent stable tag for a given major and minor.
+     *
+     * @param int|null $minor If null, get the most recent minor for the given major.
+     */
+    private function discoverMostRecentTag(int $major, ?int $minor, array $tags): string
```

Comment:
Seems like we should have some unit testing of some kind for this

Reply:
The unit tests haven't been updated in 2 years.... How concerned are we about test coverage for cow?

Reply:
not hugely concerned, though this particular method has a pair on ints and an array for parameters, so it should be very easy to unit test this

Reply:
Fair. Done.